kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/int-simplification.k

```diff
@@ -151,11 +155,15 @@ module INT-SIMPLIFICATION-COMMON
     rule A /Int 1 => A  [simplification, preserves-definedness]
 
     // safeMath mul check c / a == b where c == a * b
-    rule (A *Int B) /Int A => B requires A =/=Int 0 [simplification, preserves-definedness]
+    rule (A *Int B) /Int A => B requires notBool A ==Int 0 [simplification, preserves-definedness]
+    rule (B *Int A) /Int A => B requires notBool A ==Int 0 [simplification, preserves-definedness]
 
-    rule ((A *Int B) /Int C) /Int B => A /Int C requires B =/=Int 0 andBool C =/=Int 0 [simplification, preserves-definedness]
+    rule ((A *Int B) /Int C) /Int B => A /Int C requires 0 <Int B andBool 0 <Int C [simplification, preserves-definedness]
+    rule ((A *Int B) /Int C) /Int A => B /Int C requires 0 <Int A andBool 0 <Int C [simplification, preserves-definedness]
 
-    rule (A /Int 32) *Int 32 => A requires A modInt 32 ==Int 0 [simplification, preserves-definedness]
+    rule (A *Int B) /Int C => B *Int (A /Int C)
+      requires notBool C ==Int 0 andBool A modInt C ==Int 0
+      [simplification, concrete(A, C), preserves-definedness]
```

Comment:
Just an observation, but maybe important for lemma efficiency? You'd know better than me at this point what makes the backend slow/fast. It looks to me that the first and last lemmas here are redundant:

```
    rule (A *Int B) /Int A => B requires notBool A ==Int 0 [simplification, preserves-definedness]
    rule (B *Int A) /Int A => B requires notBool A ==Int 0 [simplification, preserves-definedness]

    rule (A *Int B) /Int C => B *Int (A /Int C)
      requires notBool C ==Int 0 andBool A modInt C ==Int 0
      [simplification, concrete(A, C), preserves-definedness]
```

It seems that if we duplicated the 3rd rule (to handle the `(B *Int A) /Int C` case, then the first two lemmas would be fully redundant. They would basically reduce to `(A *Int B) /Int A => B *Int (A /Int A)` and `(B *Int A) /Int A => B *Int (A /Int A)`.

Reply:
Yes yes, that's correct, let me take care of that.

Reply:
Ah, the third one has `concrete(A, C)`, I can't rembmer if there was a specific reason for that...

Reply:
So, I think that the reason I put `concrete(A, C)` there is that I didn't want the backend to even think about symbolic `A modInt C ==Int 0`, especially given that it will now call Z3 to figure that out. The lemmas are redundant, but I think it's better for the backend to have them all.